internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala

```diff
@@ -340,17 +340,17 @@ object Incremental {
       incremental.invalidateInitial(previous.relations, initialChanges)
 
     // If there's any compilation at all, invalidate all java sources too, so we have access to their type information.
-    val javaSources: Set[VirtualFileRef] = sources
-      .filter(_.name.endsWith(".java"))
-      .map(_.asInstanceOf[VirtualFileRef])
+    val javaSources: Set[VirtualFileRef] = sources.collect {
+      case s: VirtualFileRef if s.name.endsWith(".java") => s
+    }
     val isPickleJava = currentSetup.options.scalacOptions.contains("-Ypickle-java")
     assert(
       javaSources.isEmpty || !options.pipelining || isPickleJava,
       s"-Ypickle-java must be included into scalacOptions if pipelining is enabled with Java sources"
     )
     val initialInvSources =
-      if (initialInvSources0.nonEmpty) initialInvSources0 ++ javaSources
-      else Set.empty[VirtualFileRef]
+      if (isPickleJava && initialInvSources0.nonEmpty) initialInvSources0 ++ javaSources
+      else initialInvSources0
```

Comment:
👍  There's no point in invalidating Java sources unless `isPickleJava` is on.

Reply:
Cool. I was about to submit a comment justifying this choice but I don't need to since you've confirmed my intuition.